File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        set<ListNode*> st;
        
        while(headA!=NULL || headB!=NULL) {
            if(headA!=NULL && st.find(headA) != st.end()) {
                return headA;
            }
            
            if(headA!=NULL)  {
                st.insert(headA);
                headA = headA->next;
            }
            
            if(headB!=NULL && st.find(headB) != st.end()) {
                return headB;
            }
            
            if(headB!=NULL) {
                st.insert(headB);
                headB = headB->next;
            }
        }
        
        return NULL;
    }
};

File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, A: ListNode, B: ListNode) -> Optional[ListNode]:
        first_list_map = {}
        temp = A
        while temp:
            first_list_map[temp] = 1
            temp = temp.next

        temp = B
        while temp:
            if temp in first_list_map:
                return temp
            temp = temp.next
        return None